Mathlib/RingTheory/DedekindDomain/Factorization.lean

```diff
@@ -177,3 +223,289 @@ theorem finprod_heightOneSpectrum_factorization_coe (I : Ideal R) (hI : I ≠ 0)
 #align ideal.finprod_height_one_spectrum_factorization_coe Ideal.finprod_heightOneSpectrum_factorization_coe
 
 end Ideal
+
+/-! ### Factorization of fractional ideals of Dedekind domains -/
+
+namespace FractionalIdeal
+
+open Int IsLocalization
+
+/-- If `I` is a nonzero fractional ideal, `a ∈ R`, and `J` is an ideal of `R` such that
+`I = a⁻¹J`, then `I` is equal to the product `∏_v v^(val_v(J) - val_v(a))`. -/
+theorem finprod_heightOneSpectrum_factorization {I : FractionalIdeal (nonZeroDivisors R) K}
+    (hI : I ≠ 0) {a : R} {J : Ideal R}
+    (haJ : I = spanSingleton (nonZeroDivisors R) ((algebraMap R K) a)⁻¹ * ↑J) :
+    ∏ᶠ v : HeightOneSpectrum R, (v.asIdeal : FractionalIdeal (nonZeroDivisors R) K) ^
+      ((Associates.mk v.asIdeal).count (Associates.mk J).factors -
+        (Associates.mk v.asIdeal).count (Associates.mk (Ideal.span {a})).factors : ℤ) = I := by
+  have hJ_ne_zero : J ≠ 0 := ideal_factor_ne_zero hI haJ
+  have hJ := Ideal.finprod_heightOneSpectrum_factorization_coe K hJ_ne_zero
+  have ha_ne_zero : Ideal.span {a} ≠ 0 := constant_factor_ne_zero hI haJ
+  have ha := Ideal.finprod_heightOneSpectrum_factorization_coe K ha_ne_zero
+  rw [haJ, ← div_spanSingleton, div_eq_mul_inv, ← coeIdeal_span_singleton, ← hJ, ← ha,
+    ← finprod_inv_distrib]
+  simp_rw [← zpow_neg]
+  rw [← finprod_mul_distrib (Ideal.finite_mulSupport_coe hJ_ne_zero)
+    (Ideal.finite_mulSupport_inv ha_ne_zero)]
+  apply finprod_congr
+  intro v
+  rw [← zpow_add₀ ((@coeIdeal_ne_zero R _ K _ _ _ _).mpr v.ne_bot), sub_eq_add_neg]
+
+/-- For a nonzero `k = r/s ∈ K`, the fractional ideal `(k)` is equal to the product
+`∏_v v^(val_v(r) - val_v(s))`. -/
+theorem finprod_heightOneSpectrum_factorization_principal
```

Comment:
I think this should take `r` and `s` as arguments, and show that `... = spanSingleton R⁰ (r/s)`

Reply:
I have added the version you propose, but I need to keep mine - I need a version that I can apply to `k : K`, without needing to express it as a fraction.